### PR TITLE
Web: Add pre-relabel labels to status page.

### PR DESCRIPTION
**Pull request by brian-brazil**
Figuring out what's going on with the new service discovery
and labels is difficult. Add a popover with the labels
to the target table to make things simpler, and help
discovery of potentially useful labels.

@fabxc

### CODE REVIEW

**Comment by fabxc** [comment]
I'd like to keep the target implementation agnostic of things such as relabeling. This is the condition by which the labels are removed before a target is created, which is also agnostic of the relabeling stage:

```
if strings.HasPrefix(string(ln), clientmodel.MetaLabelPrefix) {
    delete(labels, ln)
}
```
So renaming things from `preRelabelLabels` to `metaLabels` would be consistent with that and more generic.

:+1: in general. That's a good idea for debugging.

File: retrieval/target.go [hunk 1]
```diff
@@ -156,6 +156,8 @@ type Target struct {
 	sync.RWMutex
 	// url is the URL to be scraped. Its host is immutable.
 	url *url.URL
+	// Any base labels before any relabelling.
```
Comment (fabxc) [inline retrieval/target.go]: We eventually agreed to use `relabeling` everywhere.

File: retrieval/target.go [hunk 2]
```diff
@@ -185,7 +187,7 @@ func (t *Target) Status() *TargetStatus {
 
 // Update overwrites settings in the target that are derived from the job config
 // it belongs to.
-func (t *Target) Update(cfg *config.ScrapeConfig, baseLabels clientmodel.LabelSet) {
+func (t *Target) Update(cfg *config.ScrapeConfig, baseLabels clientmodel.LabelSet, preRelabelLabels clientmodel.LabelSet) {
```
Comment (fabxc) [inline retrieval/target.go]: `a T, b T` can be shortened to `a, b T`.

File: retrieval/target.go [hunk 3]
```diff
@@ -400,6 +403,17 @@ func (t *Target) BaseLabels() clientmodel.LabelSet {
 	return lset
 }
 
+// BaseLabels returns a copy of the target's base labels before relabelling.
```
Comment (fabxc) [inline retrieval/target.go]: Comment doesn't match.

**Comment by brian-brazil** [comment]
Changed updated per comments, PTAL.

File: retrieval/target.go [hunk 4]
```diff
@@ -185,7 +187,7 @@ func (t *Target) Status() *TargetStatus {
 
 // Update overwrites settings in the target that are derived from the job config
 // it belongs to.
-func (t *Target) Update(cfg *config.ScrapeConfig, baseLabels clientmodel.LabelSet) {
+func (t *Target) Update(cfg *config.ScrapeConfig, baseLabels clientmodel.LabelSet, metaLabels clientmodel.LabelSet) {
```
Comment (fabxc) [inline retrieval/target.go]: `baseLabels, metaLabels clientmodel.LabelSet`

Comment (brian-brazil) [inline retrieval/target.go]: Done

**Comment by fabxc** [comment]
:+1:

File: web/templates/_base.html [hunk 5]
```diff
@@ -4,11 +4,17 @@
     <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
     <title>Prometheus Time Series Collection and Processing Server</title>
     <script src="{{ pathPrefix }}/static/vendor/js/jquery.min.js"></script>
+    <script src="{{ pathPrefix }}/static/vendor/bootstrap-3.3.1/js/bootstrap.min.js"></script>
```
Comment (juliusv) [inline web/templates/_base.html]: If we include this in `_base.html`, it should not be included again from `graph.html`.

File: web/templates/status.html [hunk 6]
```diff
@@ -56,7 +56,9 @@
                 </span>
               </td>
               <td>
-                {{stripLabels .BaseLabels "job" "instance"}}
+                <a href="#" data-toggle="tooltip" title="" data-original-title="Before Relabel: {{.MetaLabels}}">
```
Comment (juliusv) [inline web/templates/status.html]: Hmm, with long DNS SRV names and such, the labels kind of overflow the tooltip and aren't completely visible. Over time, we should find a better visualization here.

Comment (brian-brazil) [inline web/templates/status.html]: Yeah, this should go on a separate page but there was a fair bit of plumbing involved for that to work so I went with something simple to start with.

**Pull request merged**